File: src/main/kotlin/customfit/ai/kotlinclient/core/config/MutableCFConfig.kt
package customfit.ai.kotlinclient.core.config

import customfit.ai.kotlinclient.logging.Timber
import customfit.ai.kotlinclient.core.config.CFConfig
import java.util.concurrent.CopyOnWriteArrayList
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** A wrapper around CFConfig that allows for dynamic updates to configuration values at runtime. */
class MutableCFConfig(initConfig: CFConfig) {

    private val mutex = Mutex()
    private var _config = initConfig

    /** Get the current immutable configuration */
    val config: CFConfig
        get() = _config

    /** Listeners that will be notified when config values change */
    private val configChangeListeners = CopyOnWriteArrayList<ConfigChangeListener>()

    // Delegate properties for common access
    val clientKey: String
        get() = _config.clientKey
    val dimensionId: String?
        get() = _config.dimensionId
    val offlineMode: Boolean
        get() = _config.offlineMode
    val eventsQueueSize: Int
        get() = _config.eventsQueueSize
    val eventsFlushTimeSeconds: Int
        get() = _config.eventsFlushTimeSeconds
    val eventsFlushIntervalMs: Long
        get() = _config.eventsFlushIntervalMs
    val summariesQueueSize: Int
        get() = _config.summariesQueueSize
    val summariesFlushTimeSeconds: Int
        get() = _config.summariesFlushTimeSeconds
    val summariesFlushIntervalMs: Long
        get() = _config.summariesFlushIntervalMs
    val sdkSettingsCheckIntervalMs: Long
        get() = _config.sdkSettingsCheckIntervalMs
    val networkConnectionTimeoutMs: Int
        get() = _config.networkConnectionTimeoutMs
    val networkReadTimeoutMs: Int
        get() = _config.networkReadTimeoutMs
    val loggingEnabled: Boolean
        get() = _config.loggingEnabled
    val debugLoggingEnabled: Boolean
        get() = _config.debugLoggingEnabled
    val autoEnvAttributesEnabled: Boolean
        get() = _config.autoEnvAttributesEnabled

    // Background polling settings
    val disableBackgroundPolling: Boolean
        get() = _config.disableBackgroundPolling
    val backgroundPollingIntervalMs: Long
        get() = _config.backgroundPollingIntervalMs
    val useReducedPollingWhenBatteryLow: Boolean
        get() = _config.useReducedPollingWhenBatteryLow
    val reducedPollingIntervalMs: Long
        get() = _config.reducedPollingIntervalMs
    val maxStoredEvents: Int
        get() = _config.maxStoredEvents

    /**
     * Updates the offline mode setting
     *
     * @param enabled true to enable offline mode, false to disable
     */
    suspend fun setOfflineMode(enabled: Boolean) = updateConfig {
        _config.copy(offlineMode = enabled)
    }

    /**
     * Updates the events flush interval
     *
     * @param intervalMs the new interval in milliseconds
     */
    suspend fun setEventsFlushIntervalMs(intervalMs: Long) = updateConfig {
        require(intervalMs > 0) { "Interval must be greater than 0" }
        _config.copy(eventsFlushIntervalMs = intervalMs)
    }

    /**
     * Updates the summaries flush interval
     *
     * @param intervalMs the new interval in milliseconds
     */
    suspend fun setSummariesFlushIntervalMs(intervalMs: Long) = updateConfig {
        require(intervalMs > 0) { "Interval must be greater than 0" }
        _config.copy(summariesFlushIntervalMs = intervalMs)
    }

    /**
     * Updates the SDK settings check interval
     *
     * @param intervalMs the new interval in milliseconds
     */
    suspend fun setSdkSettingsCheckIntervalMs(intervalMs: Long) = updateConfig {
        require(intervalMs > 0) { "Interval must be greater than 0" }
        _config.copy(sdkSettingsCheckIntervalMs = intervalMs)
    }

    /**
     * Updates the network connection timeout
     *
     * @param timeoutMs the new timeout in milliseconds
     */
    suspend fun setNetworkConnectionTimeoutMs(timeoutMs: Int) = updateConfig {
        require(timeoutMs > 0) { "Timeout must be greater than 0" }
        _config.copy(networkConnectionTimeoutMs = timeoutMs)
    }

    /**
     * Updates the network read timeout
     *
     * @param timeoutMs the new timeout in milliseconds
     */
    suspend fun setNetworkReadTimeoutMs(timeoutMs: Int) = updateConfig {
        require(timeoutMs > 0) { "Timeout must be greater than 0" }
        _config.copy(networkReadTimeoutMs = timeoutMs)
    }

    /**
     * Updates the logging enabled setting
     *
     * @param enabled true to enable logging, false to disable
     */
    suspend fun setLoggingEnabled(enabled: Boolean) = updateConfig {
        _config.copy(loggingEnabled = enabled)
    }

    /**
     * Updates the debug logging enabled setting
     *
     * @param enabled true to enable debug logging, false to disable
     */
    suspend fun setDebugLoggingEnabled(enabled: Boolean) = updateConfig {
        _config.copy(debugLoggingEnabled = enabled)
    }

    /**
     * Updates the background polling enabled setting
     *
     * @param disable true to disable background polling, false to enable
     */
    suspend fun setDisableBackgroundPolling(disable: Boolean) = updateConfig {
        _config.copy(disableBackgroundPolling = disable)
    }

    /**
     * Updates the background polling interval
     *
     * @param intervalMs the new interval in milliseconds
     */
    suspend fun setBackgroundPollingIntervalMs(intervalMs: Long) = updateConfig {
        require(intervalMs > 0) { "Interval must be greater than 0" }
        _config.copy(backgroundPollingIntervalMs = intervalMs)
    }

    /**
     * Updates whether to use reduced polling when battery is low
     *
     * @param useReduced true to reduce polling on low battery
     */
    suspend fun setUseReducedPollingWhenBatteryLow(useReduced: Boolean) = updateConfig {
        _config.copy(useReducedPollingWhenBatteryLow = useReduced)
    }

    /**
     * Updates the reduced polling interval used when battery is low
     *
     * @param intervalMs the new interval in milliseconds
     */
    suspend fun setReducedPollingIntervalMs(intervalMs: Long) = updateConfig {
        require(intervalMs > 0) { "Interval must be greater than 0" }
        _config.copy(reducedPollingIntervalMs = intervalMs)
    }

    /**
     * Updates the maximum number of events stored when offline
     *
     * @param maxEvents the maximum number of events to store
     */
    suspend fun setMaxStoredEvents(maxEvents: Int) = updateConfig {
        require(maxEvents > 0) { "Max stored events must be greater than 0" }
        _config.copy(maxStoredEvents = maxEvents)
    }

    /**
     * Updates the auto environment attributes enabled setting
     *
     * @param enabled true to enable automatic environment attribute collection
     */
    suspend fun setAutoEnvAttributesEnabled(enabled: Boolean) = updateConfig {
        _config.copy(autoEnvAttributesEnabled = enabled)
    }

    /**
     * Generic method to update the config with a new value
     *
     * @param updateFn a function that returns the updated config
     */
    private suspend fun updateConfig(updateFn: () -> CFConfig) {
        mutex.withLock {
            val oldConfig = _config
            val newConfig = updateFn()
            _config = newConfig

            // Notify listeners
            notifyListeners(oldConfig, newConfig)
        }
    }

    /**
     * Add a listener that will be notified when config values change
     *
     * @param listener the listener to add
     */
    fun addConfigChangeListener(listener: ConfigChangeListener) {
        configChangeListeners.add(listener)
    }

    /**
     * Remove a previously added config change listener
     *
     * @param listener the listener to remove
     */
    fun removeConfigChangeListener(listener: ConfigChangeListener) {
        configChangeListeners.remove(listener)
    }

    /**
     * Notify all listeners of a config change
     *
     * @param oldConfig the previous config
     * @param newConfig the new config
     */
    private fun notifyListeners(oldConfig: CFConfig, newConfig: CFConfig) {
        for (listener in configChangeListeners) {
            try {
                listener.onConfigChanged(oldConfig, newConfig)
            } catch (e: Exception) {
                Timber.e(e, "Error notifying config change listener: ${e.message}")
            }
        }
    }

    /** Interface for objects that will be notified when config values change */
    interface ConfigChangeListener {
        /**
         * Called when a configuration value changes
         *
         * @param oldConfig the previous config
         * @param newConfig the new config
         */
        fun onConfigChanged(oldConfig: CFConfig, newConfig: CFConfig)
    }
}

File: customfit-reactnative-client-sdk/src/config/core/MutableCFConfig.ts
import { CFConfig } from '../../core/types/CFTypes';
import { CFConfigImpl } from './CFConfig';
import { Logger } from '../../logging/Logger';

/**
 * Mutable configuration wrapper for runtime updates
 */
export class MutableCFConfig {
  private _config: CFConfig;
  private readonly _listeners: Array<(config: CFConfig) => void> = [];

  constructor(config: CFConfig) {
    this._config = config;
  }

  /**
   * Get current immutable config
   */
  get config(): CFConfig {
    return this._config;
  }

  /**
   * Add a listener for config changes
   */
  addListener(listener: (config: CFConfig) => void): void {
    this._listeners.push(listener);
  }

  /**
   * Remove a listener
   */
  removeListener(listener: (config: CFConfig) => void): void {
    const index = this._listeners.indexOf(listener);
    if (index > -1) {
      this._listeners.splice(index, 1);
    }
  }

  /**
   * Update configuration and notify listeners
   */
  private updateConfig(newConfig: CFConfig): void {
    this._config = newConfig;
    for (const listener of this._listeners) {
      try {
        listener(newConfig);
      } catch (e) {
        Logger.error(`Error notifying config listener: ${e}`);
      }
    }
  }

  /**
   * Create a new config with updated values
   */
  private copyWith(updates: Partial<CFConfig>): CFConfig {
    return new CFConfigImpl(
      updates.clientKey ?? this._config.clientKey,
      updates.eventsQueueSize ?? this._config.eventsQueueSize,
      updates.eventsFlushTimeSeconds ?? this._config.eventsFlushTimeSeconds,
      updates.eventsFlushIntervalMs ?? this._config.eventsFlushIntervalMs,
      updates.maxStoredEvents ?? this._config.maxStoredEvents,
      updates.maxRetryAttempts ?? this._config.maxRetryAttempts,
      updates.retryInitialDelayMs ?? this._config.retryInitialDelayMs,
      updates.retryMaxDelayMs ?? this._config.retryMaxDelayMs,
      updates.retryBackoffMultiplier ?? this._config.retryBackoffMultiplier,
      updates.summariesQueueSize ?? this._config.summariesQueueSize,
      updates.summariesFlushTimeSeconds ?? this._config.summariesFlushTimeSeconds,
      updates.summariesFlushIntervalMs ?? this._config.summariesFlushIntervalMs,
      updates.sdkSettingsCheckIntervalMs ?? this._config.sdkSettingsCheckIntervalMs,
      updates.networkConnectionTimeoutMs ?? this._config.networkConnectionTimeoutMs,
      updates.networkReadTimeoutMs ?? this._config.networkReadTimeoutMs,
      updates.loggingEnabled ?? this._config.loggingEnabled,
      updates.debugLoggingEnabled ?? this._config.debugLoggingEnabled,
      updates.logLevel ?? this._config.logLevel,
      updates.offlineMode ?? this._config.offlineMode,
      updates.disableBackgroundPolling ?? this._config.disableBackgroundPolling,
      updates.backgroundPollingIntervalMs ?? this._config.backgroundPollingIntervalMs,
      updates.useReducedPollingWhenBatteryLow ?? this._config.useReducedPollingWhenBatteryLow,
      updates.reducedPollingIntervalMs ?? this._config.reducedPollingIntervalMs,
      updates.autoEnvAttributesEnabled ?? this._config.autoEnvAttributesEnabled
    );
  }

  /**
   * Update SDK settings check interval
   */
  updateSdkSettingsCheckInterval(intervalMs: number): void {
    if (intervalMs <= 0) {
      throw new Error('SDK settings check interval must be greater than 0');
    }
    this.updateConfig(this.copyWith({ sdkSettingsCheckIntervalMs: intervalMs }));
  }

  /**
   * Update events flush interval
   */
  updateEventsFlushInterval(intervalMs: number): void {
    if (intervalMs <= 0) {
      throw new Error('Events flush interval must be greater than 0');
    }
    this.updateConfig(this.copyWith({ eventsFlushIntervalMs: intervalMs }));
  }

  /**
   * Update summaries flush interval
   */
  updateSummariesFlushInterval(intervalMs: number): void {
    if (intervalMs <= 0) {
      throw new Error('Summaries flush interval must be greater than 0');
    }
    this.updateConfig(this.copyWith({ summariesFlushIntervalMs: intervalMs }));
  }

  /**
   * Update network connection timeout
   */
  updateNetworkConnectionTimeout(timeoutMs: number): void {
    if (timeoutMs <= 0) {
      throw new Error('Network connection timeout must be greater than 0');
    }
    this.updateConfig(this.copyWith({ networkConnectionTimeoutMs: timeoutMs }));
  }

  /**
   * Update network read timeout
   */
  updateNetworkReadTimeout(timeoutMs: number): void {
    if (timeoutMs <= 0) {
      throw new Error('Network read timeout must be greater than 0');
    }
    this.updateConfig(this.copyWith({ networkReadTimeoutMs: timeoutMs }));
  }

  /**
   * Set debug logging enabled
   */
  setDebugLoggingEnabled(enabled: boolean): void {
    this.updateConfig(this.copyWith({ debugLoggingEnabled: enabled }));
  }

  /**
   * Set logging enabled
   */
  setLoggingEnabled(enabled: boolean): void {
    this.updateConfig(this.copyWith({ loggingEnabled: enabled }));
  }

  /**
   * Set offline mode
   */
  setOfflineMode(offline: boolean): void {
    this.updateConfig(this.copyWith({ offlineMode: offline }));
  }
} 